### PR TITLE
Do some detection of messed up installs and some remediation

### PR DESCRIPTION
**Pull request by zanderz**
Earlier experimental versions actually ran uninstall, but that is too dangerous to do automatically IMHO. Let's start by shipping this.
@keybase/react-hackers

### CODE REVIEW

File: keybase/platform_windows.go [hunk 1]
```diff
@@ -211,10 +212,149 @@ func (c context) PausedPrompt() bool {
...
+
+	UIDs, err := k.ReadSubKeyNames(-1)
+	if err != nil {
+		c.log.Infof("Error reading subkeys: %s\n", err.Error())
+		return
+	}
+	for _, UID := range UIDs {
+		componentsKey, err := registry.OpenKey(k, UID+"\\Components", readAccess)
+		if err != nil {
+			c.log.Infof("Error opening subkey %s: %s\n", UID+"\\Components", err.Error())
```
Comment (maxtaco) [inline keybase/platform_windows.go]: This error is just dropped and we charge right through. Is that OK?

File: keybase/platform_windows.go [hunk 2]
```diff
@@ -211,10 +212,149 @@ func (c context) PausedPrompt() bool {
...
+			if err != nil {
+				if componentKeyAccess == writeAccess {
+					c.log.Infof("Error opening subkey for writing %s: %s\n", componentKeyName, err.Error())
+					componentKeyAccess = readAccess
+					componentKey, err = registry.OpenKey(componentsKey, componentKeyName, componentKeyAccess)
+				}
+				// TODO: No need to list all the components we couldn't open in write mode.
+				// This is expected when run without elevated permissions.
+				if err != nil {
+					continue
```
Comment (maxtaco) [inline keybase/platform_windows.go]: does it make sense to warn in the case of this error?

File: keybase/platform_windows.go [hunk 3]
```diff
@@ -211,10 +212,149 @@ func (c context) PausedPrompt() bool {
...
+						}
+					}
+					foundKeybase = true
+				}
+			}
+			componentKey.Close()
+		}
+		componentsKey.Close()
+	}
+	return
```
Comment (maxtaco) [inline keybase/platform_windows.go]: Change to `return result`

File: keybase/platform_windows.go [hunk 4]
```diff
@@ -211,10 +212,149 @@ func (c context) PausedPrompt() bool {
...
+			c.log.Infof("Removing %s", f)
+			if err = os.Remove(f); err != nil {
+				c.log.Infof("Error removing file: %s", err.Error())
+			}
+		}
+	}
+}
+
+// DeepClean is called when a faulty upgrade has been detected
+// Eventually we may need to preform full uninstalls but that is kind of riskyk
```
Comment (maxtaco) [inline keybase/platform_windows.go]: `riskyk` misspelling

File: keybase/platform_windows.go [hunk 5]
```diff
@@ -211,10 +212,149 @@ func (c context) PausedPrompt() bool {
 	return false
 }
 
+// checkRegistryComponents will fall back to read-only if it can't do deletion,
+// returning true if either entries were deleted, or multiple Keybase products were found
+func (c context) checkRegistryComponents(wow64, readonly bool) (result bool) {
```
Comment (maxtaco) [inline keybase/platform_windows.go]: `readonly` is only ever set to `true`.  Can you get rid of it?

Comment (maxtaco) [reply keybase/platform_windows.go]: It's bugprone to have functions take a long list of boolean arguments.

File: keybase/platform_windows.go (hunk 1 above)
Comment (zanderz) [reply keybase/platform_windows.go]: Don't think so!

**Comment by maxtaco** [comment]
@zanderz just happened across this tab. LMK when it's ready for another look

**Comment by zanderz** [comment]
yes I think it's good for another look now, thanks

**Pull request merged**